### DISCUSSION TITLE
Page numbers not storing and dynamic fields

### DISCUSSION THREAD

**Discussion by genzerstech** [milvus-io/milvus#30290] [Q&A and General discussion]
We are using langchain to store embeddings from pdf files. when we print the documents on console, we can see the associated page numbers of the chunks. however, when we use attu to look at the milvus collection, the page number is not showing. 

We can see an empty dict kind of object under the dynamic fields column. 

We are then using RetrievalQA of langchain to retrieve the results. it retrieves the responses fine and also has the associated correct source document for the response. however, we are not seeing any associated page numbers. Same code works with ChromaDB and we are able to see the page numbers. 

2 questions
1) Does Milvus save the page numbers in its metadata?
2) What is the use of 'Dynamic Fields' and is it possible to persist any data in it using langchain methods?

![image](https://github.com/milvus-io/milvus/assets/128766079/ecf63fc8-2735-4fdd-821e-64c1768cd2c9)

**Comment by yhmo** [comment]
'Dynamic Fields' is not applied in the implementation of Langchain.Milvus. But it can use a JSON field to store all the metadata. You can input page numbers in the metadata so that the page numbers are stored in this field.

In Langchain, you call `VectorStore.from_documents()` to add documents into vector db.
The definition of `from_document()`:
```
def from_documents(
        cls: Type[VST],
        documents: List[Document],
        embedding: Embeddings,
        **kwargs: Any,
    )
```
The definition of `Document`:
```
class Document(Serializable):
    page_content: str
    metadata: dict = Field(default_factory=dict)
```

The Milvus vector store class is derived from VectorStore, the `from_document()` passes inputs to `Milvus.from_texts()`.
The definition of `Milvus.from_texts()`:
```
def from_texts(
        cls,
        texts: List[str],
        embedding: Embeddings,
        metadatas: Optional[List[dict]] = None,
        collection_name: str = "LangChainCollection",
        connection_args: dict[str, Any] = DEFAULT_MILVUS_CONNECTION,
        consistency_level: str = "Session",
        index_params: Optional[dict] = None,
        search_params: Optional[dict] = None,
        drop_old: bool = False,
        **kwargs: Any,
    )
```

The `from_texts()` passes inputs to `add_texts()`.
The definition of `Milvus.add_texts()`:
```
def add_texts(
        self,
        texts: Iterable[str],
        metadatas: Optional[List[dict]] = None,
        timeout: Optional[int] = None,
        batch_size: int = 1000,
        **kwargs: Any,
    )
```

The from_texts() does the following tasks:
1. generate an embedding for each text
2. construct a `insert_dict: dict[str, list]` for insertion
3. call collection.insert(insert_dict) to insert the data to milvus collection

The key logic is here: https://github.com/langchain-ai/langchain/blob/08d3fd7f2e5aec466abb62171e69959f3594771e/libs/community/langchain_community/vectorstores/milvus.py#L527

If you have defined a "metadata" field, all the metadata will be put into a JSON field named `self._metadata_fiel`. Otherwise, add_texts() will treat each key-value pair as a separate field.

The collection schema is defined when you initialize the Milvus object.
We can see the constructor has a parameter "metadata_field". The default value is Null.
```
class Milvus(VectorStore):
  def __init__(
        self,
        embedding_function: Embeddings,
        collection_name: str = "LangChainCollection",
        collection_description: str = "",
        collection_properties: Optional[dict[str, Any]] = None,
        connection_args: Optional[dict[str, Any]] = None,
        consistency_level: str = "Session",
        index_params: Optional[dict] = None,
        search_params: Optional[dict] = None,
        drop_old: Optional[bool] = False,
        *,
        primary_field: str = "pk",
        text_field: str = "text",
        vector_field: str = "vector",
        metadata_field: Optional[str] = None,
        partition_key_field: Optional[str] = None,
        partition_names: Optional[list] = None,
        replica_number: int = 1,
        timeout: Optional[float] = None,
    ):
```

The `__init__()` calls `_create_collection()` to create a collection: https://github.com/langchain-ai/langchain/blob/08d3fd7f2e5aec466abb62171e69959f3594771e/libs/community/langchain_community/vectorstores/milvus.py#L298

In the `_create_collection()` we can see if you have provided the "metadata_field", it will use this name to declare a JSON field to store all metadata data. Otherwise, it will deduce some separate fields when you first time call the from_document(), according to the "metadatas" you provided.


For example:
We declare a Milvus object with metadata_field = "properties". (or use from_documents() to construct a Milvus object also ok, pass the "metadata_field" by the "**kwargs")

```
store = Milvus(
  ......
  metadata_field = "properties",
)
```

Then we call from_documents() to add documents:

```
doc1 = Document(page_content="aaaaa", metadata = {"year": 1986, "published": true, "source": "http1", "page_num": 32332})
doc2 = Document(page_content="bbbbbb", metadata = {"author": "Bob", "published": false, "source": "http2", "page_num": 654})

store.from_documents(
    documents: List[doc1, doc2],
    embedding: Embeddings,
)
```

From the attu, we will see the collection has 4 fields: "pk", "text", "vector", "properties".
All the content of metadata is stored in "properties" field. And you will see the "page_num" in this field.

**Comment by genzerstech** [comment]
This is great.. thank you so much. Answers so many questions. 

However, we realized that inserting the data directly using methods form pyMilvus module is better approach for us and this is what we have been using. 

So question is actually not related to Langchain. It seems more of an issue with pymilvus itself. 

I am using the below to insert documents. 

            Milvus.from_texts(collection_name=collection_name, texts=[text.page_content], embedding=embeddings,
                connection_args = {"host": os.environ.get("MILVUS_HOST"), "port": os.environ.get("MILVUS_PORT"),"user":os.environ.get("MILVUS_USER"),"password":os.environ.get("MILVUS_PASS")},
                metadatas=[{"source":"adf","asdf":"asdfaa","dferg":"asdfadfff","dsgs4":"agv"}])



When retrieving the data, we are getting only 'source' in the metadata and none of the other metadata that I am randomly inserting above. What am I doing wrong? 


![image](https://github.com/milvus-io/milvus/assets/128766079/7de5802e-ca25-4b06-8021-9244b14bc38e)


We are using langchain to retrieve the data. perhaps the metadata exists but is it somehow fetching only the 'source' key of the metadata dictionary. is there a way to retrieve the entire metadata?

**Reply by yhmo** [reply]
As I mentioned in the comment, use a "metadata_field" to define a JSON field to store all the metadata. Otherwise it will infer the collection schema by the first item.

```
from tests.integration_tests.vectorstores.fake_embeddings import (
    FakeEmbeddings,
)

from langchain_community.vectorstores import Milvus

if __name__ == '__main__':

    milvus = Milvus.from_texts(
        connection_args={"host": "localhost", "port": 19530},
        collection_name="test_fake",
        texts=["hello earth", "monday morning", "world is ok"],
        embedding=FakeEmbeddings(),
        metadata_field="meta",                                # this parameter defines a JSON to store all the metadatas
        metadatas=[
            {"source": "adf", "asdf": "asdfaa", "dferg": "asdfadfff", "dsgs4": "agv"},
            {"source": "vf", "vff": "vfff", "dferg": "vffff", "vfff": "ffv"},
            {"source": "uy", "uyuy": "uyuy", "uyyu": "uyuy", "uyyy": "uyuy"},
        ])

    docs = milvus.similarity_search("hello world",k=5)
    print("==========search result===========")
    for doc in docs:
        print(doc)
    print("==================================")

```

The results:
```
==========search result===========
page_content='hello earth' metadata={'source': 'adf', 'asdf': 'asdfaa', 'dferg': 'asdfadfff', 'dsgs4': 'agv'}
page_content='monday morning' metadata={'source': 'vf', 'vff': 'vfff', 'dferg': 'vffff', 'vfff': 'ffv'}
page_content='world is ok' metadata={'source': 'uy', 'uyuy': 'uyuy', 'uyyu': 'uyuy', 'uyyy': 'uyuy'}
==================================

Process finished with exit code 0
```

**Comment by genzerstech** [comment]
this is perfect. thank you so much. Looks like langchain has to be upgraded to 0.1 though in order for this to work.. which we did of course.
